File: packages/python/llm_local_client/llm_client.py
import requests
from typing import Optional, List, Dict, Any
from .context.context import Context
from .infrastructure.model_manager import ModelManager
from .infrastructure.infra_check import check_ollama_status
from .utils.utils import Utils

class LLMClient:
    def __init__(self, model: str, base_url: str = 'http://localhost:11434'):
        self.model = model
        self.base_url = base_url
        self.context_manager = Context()
        self.model_manager = ModelManager(self.base_url)
        self.utils = Utils(self)

    def initialize(self):
        if not check_ollama_status(self.base_url):
            raise RuntimeError(f"Ollama runtime not detected at {self.base_url}. Please run 'make initial'.")
        self.model_manager.ensure_model(self.model)

    def prompt(self, message: str, context_id: Optional[str] = None, images: Optional[List[str]] = None, temperature: Optional[float] = None) -> str:
        # 1. Record user message
        user_node = self.context_manager.add_node('user', message, context_id)

        # 2. Fetch full history for this branch
        full_context = self.context_manager.get_full_context(user_node.id)
        messages = []
        for node in full_context:
            msg = {"role": node.role, "content": node.content}
            if node.id == user_node.id and images:
                msg["images"] = images
            messages.append(msg)

        # 3. Call LLM
        payload = {
            "model": self.model,
            "messages": messages,
            "stream": False
        }
        if temperature is not None:
            payload["options"] = {"temperature": temperature}

        try:
            response = requests.post(f"{self.base_url}/api/chat", json=payload, timeout=60)
            response.raise_for_status()
            data = response.json()
            response_content = data.get("message", {}).get("content", "")

            # 4. Record assistant message
            self.context_manager.add_node('assistant', response_content, user_node.id)

            return response_content
        except requests.RequestException as e:
            raise RuntimeError(f"Failed to prompt LLM: {e}")

    def get_context(self) -> Context:
        return self.context_manager

File: packages/typescript/src/llmClient.ts
import axios from 'axios';
import { Context } from './context/Context';
import { ContextNode } from './context/ContextNode';
import { ModelManager } from './infrastructure/modelManager';
import { checkOllamaStatus } from './infrastructure/infraCheck';
import { Utils } from './utils/Utils';

export interface LLMClientOptions {
  model: string;
  baseUrl?: string;
}

export interface PromptOptions {
  message: string;
  contextId?: string;
  images?: string[];
  temperature?: number;
}

export class LLMClient {
  private model: string;
  private baseUrl: string;
  private contextManager: Context;
  private modelManager: ModelManager;
  public utils: Utils;

  constructor(options: LLMClientOptions) {
    this.model = options.model;
    this.baseUrl = options.baseUrl || 'http://localhost:11434';
    this.contextManager = new Context();
    this.modelManager = new ModelManager(this.baseUrl);
    this.utils = new Utils(this);
  }

  async initialize(): Promise<void> {
    const isRunning = await checkOllamaStatus(this.baseUrl);
    if (!isRunning) {
      throw new Error(`Ollama runtime not detected at ${this.baseUrl}. Please run 'make initial'.`);
    }
    await this.modelManager.ensureModel(this.model);
  }

  async prompt(options: PromptOptions): Promise<string> {
    const { message, contextId, images, temperature } = options;

    // 1. Record user message in context tree
    const userNode = this.contextManager.addNode('user', message, contextId || null);

    // 2. Fetch full history for this branch
    const fullContext = this.contextManager.getFullContext(userNode.id);
    const messages = fullContext.map(node => {
      const msg: any = { role: node.role, content: node.content };
      if (node.id === userNode.id && images && images.length > 0) {
        msg.images = images;
      }
      return msg;
    });

    // 3. Call LLM
    try {
      const response = await axios.post(`${this.baseUrl}/api/chat`, {
        model: this.model,
        messages: messages,
        stream: false,
        options: {
          temperature: temperature
        }
      });

      const responseContent = response.data.message.content;

      // 4. Record assistant message
      this.contextManager.addNode('assistant', responseContent, userNode.id);

      return responseContent;
    } catch (error: any) {
      throw new Error(`Failed to prompt LLM: ${error.message}`);
    }
  }

  getContext(): Context {
    return this.contextManager;
  }
}
